src/opt/pre_scheduling_impl.cpp

```
@@ -0,0 +1,334 @@
+            if(arg_s == stream)
+                continue;
+            if(!has_mask(arg, record_event))
+            {
+                events.push_back(event);
+                arg->set_event(event);
+                arg->add_mask(record_event);
+                if(enable_event_as_instr)
+                    insert_instr.insert_record_event(p_program, std::next(arg), event);
+                event++;
+            }
+
+            ins->add_mask(wait_event);
+            add_mask(arg, record_event);
+            add_mask(ins, wait_event);
+        }
+        if(enable_event_as_instr)
+        {
+            for(auto&& i : events)
+                insert_instr.insert_wait_event(p_program, ins, i);
+        }
+    }
+}
+
+void pre_scheduling_impl::run()
+{
+    std::size_t num_of_instrs = p_program->size();
+    if(num_of_instrs == 0)
+        return;
+    MIGRAPHX_DEBUG(dump("---Before pre-scheduling---"));
+    MIGRAPHX_DEBUG(dump_program());
+    nodes.resize(num_of_instrs);
+    compute_weights();
+    reorder();
+}
+
+#ifdef MIGRAPHX_DEBUG_OPT
+void pre_scheduling_impl::dump(const std::string& str) { std::cout << str << std::endl; }
+
+void pre_scheduling_impl::dump_program() { std::cout << *p_program << std::endl; }
+
+void pre_scheduling_impl::dump(std::list<dag_node*>& sorted_nodes)
+{
+    for(auto&& node : sorted_nodes)
+    {
+        node->dump();
+        if(!node->ins->inputs().empty())
+        {
+            std::cout << " inputs: ";
+            for(auto&& arg : node->ins->inputs())
+            {
+                dag_node* def_node = instr2_node[arg];
+                std::cout << " @" << def_node->ins_ndx;
+            }
+            std::cout << std::endl;
+        }
+    }
+}
+
+void pre_scheduling_impl::verify()
+{
+    std::unordered_map<instruction_ref, bool> visited;
+    for(auto ins : iterator_for(*p_program))
+    {
+        for(auto&& arg : ins->inputs())
+        {
+            assert(visited.find(arg) != visited.end());
```

This should throw an exception, otherwise this is never checked in non-debug builds.